gitman/git.py

```diff
@@ -16,49 +16,67 @@ def git(*args, **kwargs):
     return call('git', *args, **kwargs)
 
 
+def gitsvn(*args, **kwargs):
+    return call('git', 'svn', *args, **kwargs)
+
+
+def in_svn_repo():
+    """Determine if the current working directory is inside a subversion repository."""
+    output = git('show', '--shortstat', _show=False, _ignore=True)
+    if any('git-svn-id' in line for line in output):
+        return True
+    return False
+
+
 def clone(repo, path, *, cache=settings.CACHE, sparse_paths=None, rev=None):
     """Clone a new Git repository."""
     log.debug("Creating a new repository...")
-
-    name = repo.split('/')[-1]
-    if name.endswith(".git"):
-        name = name[:-4]
-
-    reference = os.path.join(cache, name + ".reference")
-    if not os.path.isdir(reference):
-        git('clone', '--mirror', repo, reference)
-
-    normpath = os.path.normpath(path)
-    if sparse_paths:
-        os.mkdir(normpath)
-        git('-C', normpath, 'init')
-        git('-C', normpath, 'config', 'core.sparseCheckout', 'true')
-        git('-C', normpath, 'remote', 'add', '-f', 'origin', reference)
-
-        with open("%s/%s/.git/info/sparse-checkout" % (os.getcwd(), normpath), 'w') as fd:
-            fd.writelines(sparse_paths)
-        with open("%s/%s/.git/objects/info/alternates" % (os.getcwd(), normpath), 'w') as fd:
-            fd.write("%s/objects" % reference)
-
-        # We use directly the revision requested here in order to respect,
-        # that not all repos have `master` as their default branch
-        git('-C', normpath, 'pull', 'origin', rev)
-    else:
-        git('clone', '--reference', reference, repo, os.path.normpath(path))
+    try:
+        name = repo.split('/')[-1]
+        if name.endswith(".git"):
+            name = name[:-4]
+
+        reference = os.path.join(cache, name + ".reference")
+        if not os.path.isdir(reference):
+            git('clone', '--mirror', repo, reference)
+
+        normpath = os.path.normpath(path)
+        if sparse_paths:
+            os.mkdir(normpath)
+            git('-C', normpath, 'init')
+            git('-C', normpath, 'config', 'core.sparseCheckout', 'true')
+            git('-C', normpath, 'remote', 'add', '-f', 'origin', reference)
+
+            with open("%s/%s/.git/info/sparse-checkout" % (os.getcwd(), normpath), 'w') as fd:
+                fd.writelines(sparse_paths)
+            with open("%s/%s/.git/objects/info/alternates" % (os.getcwd(), normpath), 'w') as fd:
+                fd.write("%s/objects" % reference)
+
+            # We use directly the revision requested here in order to respect,
+            # that not all repos have `master` as their default branch
+            git('-C', normpath, 'pull', 'origin', rev)
+        else:
+            git('clone', '--reference', reference, repo, os.path.normpath(path))
+    except ShellError:
```

Comment:
Can this check be applied any sooner?

Reply:
To determine that a http/https URL is a SVN or git repo is not easy to do. Therefore this trial and error approach has been choosen. Another way could be to specify the repo type in the gitman.yml.
For example:
```yml
- name: TAs
  svn-repo: http://my_subversion/trunk/my_project
  rev: HEAD
```

Or something like that.

As second option could be that gitman try to resolve remote repository information berforehand, like:

```bash
⟩ git ls-remote -h  https://github.com/lz4/lz4
179670f32f4960f00a89c7ef08382b06dd024873        refs/heads/betterBench
0ddd1ceb1db0f3a40a5231d52f59d25ceacd6480        refs/heads/check
5e7780d2d8394a5feb4fe5758c22e826b5cb689f        refs/heads/coverity_scan
6c23f03b93608a444f8747940bca26708a58fa78        refs/heads/dev
9dc249ee3cb3482a28762912132021094feb55a1        refs/heads/fasterDict
25b16e8a2e51c41f5864ce5cc94ce65ab6fd52b9        refs/heads/fasterHC
289603abdee89861ae072fb43fab79b7fd34c08c        refs/heads/gh-pages
4421db9352e705525e0d7196180536375158d587        refs/heads/lowAddr
dfed9fa1d77f0434306d377c4da1f7191d3ba08a        refs/heads/master
550b40849f6b77850453a1190b97430b121802bc        refs/heads/mergeOpt
7173a631db61ab9535bd0d6e5e00e9dc081d4df3        refs/heads/mflimit

⟩ git ls-remote -h http://ls/trunk/proj/
fatal: unable to access 'http://ls/trunk/proj/': The requested URL returned error: 502

```

Use like:

```python
def is_git_remote_repo(repo):
    """Determine if the remote repo is a git repository."""
    #git ls-remote -h https://github.com/lz4/lz4
    output = git('ls-remote -h', repo)
    return any('refs/heads/' in line for line in output)
```

Reply:
What do you prefer?

Reply:
Let's try that second approach.

Reply:
I have experimented a bit and found an approach.

The current approach is to use `git show-ref` to detect git and git-svn repos locally.
```python
    output = git('show-ref', _show=False, _ignore=True)
    if any('refs/remotes/git-svn' in line for line in output): # check if out contains refs/remotes/git-svn -> GIT_SVN
        repotype = RepoType.GIT_SVN
    elif any('refs/remotes/origin' in line for line in output): # check if output contains refs/remotes/origin -> GIT
        repotype = RepoType.GIT
```
For remote detection I use `git ls-remote` and `svn info`:

```python
  output = git('ls-remote', repo, _show=False, _ignore=True)
    if any('refs/heads' in line for line in output): # check if output contains refs/heads -> GIT
        repotype = RepoType.GIT    
    elif any('fatal' in line for line in output): # check if output contains fatal -> NOT GIT
        output = svn('info', repo, _show=False, _ignore=True)
        if any('UUID' in line for line in output): # check if output contains UUID -> SVN
            repotype = RepoType.SVN
```
But now I'm not convinced about the automatic detection of git and svn repositories. It is actually some logic that bloat up the code complexity for something that does not necessarily needs to be done automatically. It also costs some processing time to resolve the remote info.
I think it would be better to specify the repo type in the config as other buildsystems do it in a similar way (jenkins). Like:

```
- name: TAs
  class: git-svn
  repo: http://my_subversion/trunk/my_project
  rev: HEAD
```

The `class` (repo type) could be `git` by default and for the future other strategies to resolve dependencies could be implemented (e.g. `git-lfs`, `http-targz` ...).
This would simplify the code structure and development reps. testing (reduces possible side effects).

What do you think?

Reply:
That seems reasonable. Have you seen other tools call this classification "class"?